common/common.c

```diff
@@ -318,3 +318,50 @@ char **mp_dup_str_array(void *tctx, char **s)
         MP_TARRAY_APPEND(tctx, r, num_r, NULL);
     return r;
 }
+
+int mp_natural_cmp_filename(const void *a, const void *b)
+{
+    // Comparison function for an ASCII-only "natural" sort. Case is ignored and
+    // numbers are ordered by value regardless of padding. Two filenames that
+    // differ only in the padding of numbers will be considered equal and end up
+    // in arbitrary order. Bytes outside of A-Z/a-z/0-9 will by sorted by byte
+    // value.
+    char *name1 = *(char **)a, *name2 = *(char **)b;
+    while (*name1 && *name2) {
+        if (mp_isdigit(*name1) && mp_isdigit(*name2)) {
+            while (*name1 == '0')
+                name1++;
+            while (*name2 == '0')
+                name2++;
+            char *end1 = name1, *end2 = name2;
+            while (mp_isdigit(*end1))
+                end1++;
+            while (mp_isdigit(*end2))
+                end2++;
+            if ((end1 - name1) < (end2 - name2))
+                return -1;
+            if ((end1 - name1) > (end2 - name2))
+                return 1;
+            while (name1 < end1) {
+                if (*name1 < *name2)
+                    return -1;
+                if (*name1 > *name2)
+                    return 1;
+                name1++;
+                name2++;
+            }
+            continue;
+        }
+        if (mp_tolower(*name1) < mp_tolower(*name2))
+            return -1;
+        if (mp_tolower(*name1) > mp_tolower(*name2))
+            return 1;
+        name1++;
+        name2++;
```

Comment:
Do the above checks really guarantee that none of these were '\0'? This is rather hard to see from the code structure, and that's how bug happens.

Reply:
Being in the while means both `name1[0]` and `name2[0]` aren't '\0' -- that's exactly its condition. For each run through the loop body, this code is either the only thing executed, or isn't executed at all. I've replace the continue with an else as you asked. Does that make it clearer? (Or have I misunderstood your concern completely?)

Reply:
I saw that the code was correct, and my complaint was about how that was not obvious. I think the new version of the code is much better.